### PR TITLE
Homepage front

### PR DESCRIPTION
**Pull request by alejandrobodichon**
### Summary
A RecyclerView has been added with hardcode news. When pull down, news are added at the end. When have no more news, shows a Toast saying that.
### Screenshot
pr-homepage-front

### Trello card
https://trello.com/c/z14iwK1A/35-u4-home-page-frontend-news

### CODE REVIEW

File: app/src/main/java/ar/com/wolox/android/example/model/News.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package ar.com.wolox.android.example.model;
```
Comment (dylan-muszel) [inline app/src/main/java/ar/com/wolox/android/example/model/News.java]: Why Java?

File: app/src/main/java/ar/com/wolox/android/example/model/User.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package ar.com.wolox.android.example.model;
```
Comment (dylan-muszel) [inline app/src/main/java/ar/com/wolox/android/example/model/User.java]: Same as above: why Java?

File: app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/news/NewsFragment.kt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+    private lateinit var mRecyclerView: RecyclerView
+    private lateinit var viewAdapter: RecyclerView.Adapter<*>
+    private lateinit var viewManager: RecyclerView.LayoutManager
+    private lateinit var news: MutableList<News>
+
+    override fun layout(): Int = R.layout.news_fragment_page
+
+    override fun init() {
+        news = presenter.loadNews()
```
Comment (dylan-muszel) [inline app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/news/NewsFragment.kt]: In the `MVP` pattern, the `view` must not know anything about when to work. The `view` only tell events to the `presenter` and receive orders by it. So, here it's not okay. The `view` don't know when to load the news, this `List` will start empty. So, to manage with `MVP` the steps should be something like this:

- The view initialized the `news` as an empty list.
- On the `presenter.onViewAttached()`:
   - The `presenter` get the news (if them are hardcoded, skip this point).
   - If there are more news to load:
      - The `presenter` send the news to the `view` by calling a method like you have but sending the news `addNews(news: News)`.
      - The `view` add the news received to the list. 
   - If there are not more news to load:
      - The `presenter` call a method of the `view` like `showNoMoreNews()`
      - The `view` show a toast with the error message.
- On the `view` on the `setOnRefreshListener` the `view` will only inform to the presenter that the news list is refreshed and the presenter will manage the same behaviour as in the `onViewAttached`.

I know it hard to understand, so if you have any doubt, talk to me.

**Review by dylan-muszel — CHANGES_REQUESTED** [review]
Good work, but there are some important thing about `MVP` pattern that would be great to check. I also have the doubt of why do you create some classes in java.

And I found out that you joined a project. Congratulations!

File: app/build.gradle [hunk 4]
```diff
@@ -101,6 +101,17 @@ buildscript {
 dependencies {
     implementation fileTree(include: ['*.jar'], dir: 'libs')
 
+    //Fresco
+    implementation 'com.facebook.fresco:fresco:1.12.1'
+
+
```
Comment (lmiotti) [inline app/build.gradle:116]: Remove this empty line

File: app/build.gradle [hunk 5]
```diff
@@ -115,6 +126,8 @@ dependencies {
     // Dagger
     implementation "com.google.dagger:dagger:$dagger_version"
     implementation "com.google.dagger:dagger-android-support:$dagger_version"
+    implementation 'com.android.support.constraint:constraint-layout:1.1.3'
```
Comment (lmiotti) [inline app/build.gradle]: When adding new dependencies please try to add an extension for every version ass dagger is using now

File: app/src/main/AndroidManifest.xml [hunk 6]
```diff
@@ -1,32 +1,41 @@
-<manifest
-    xmlns:android="http://schemas.android.com/apk/res/android"
+<?xml version="1.0" encoding="utf-8"?>
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (lmiotti) [inline app/src/main/AndroidManifest.xml]: Send xmlns:android on a new line

File: app/src/main/AndroidManifest.xml [hunk 7]
```diff
@@ -1,32 +1,41 @@
...
         </activity>
 
         <activity
-            android:name=".example.ui.viewpager.ViewPagerActivity"
-            android:screenOrientation="portrait" />
+            android:name=".example.ui.training.signUp.SignUpActivity"
+            android:screenOrientation="portrait">
+        </activity>
 
     </application>
```
Comment (lmiotti) [inline app/src/main/AndroidManifest.xml:40]: Remove this empty line too

File: app/src/main/java/ar/com/wolox/android/example/TrainingConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package ar.com.wolox.android.example;
+
+public final class TrainingConfiguration {
+
+    private TrainingConfiguration() {
+    }
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/TrainingConfiguration.java]: If constructor is a method is empty try to put it on a single line. It's a silly thing but it makes code "easier to read"

File: app/src/main/java/ar/com/wolox/android/example/model/News.java (hunk 1 above)
Comment (lmiotti) [reply app/src/main/java/ar/com/wolox/android/example/model/News.java]: This class continue being on Java please change it to Kotlin

File: app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/ProfileFragment.kt [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+import ar.com.wolox.wolmo.core.fragment.WolmoFragment
+import ar.com.wolox.wolmo.core.presenter.BasePresenter
+import javax.inject.Inject
+
+/**
+ * LoginFragment is used for user to Login
+ */
+class ProfileFragment @Inject constructor() : WolmoFragment<BasePresenter<Any>>() {
+
+    override fun init() {
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/ProfileFragment.kt]: Same as above

File: app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/ProfileFragment.kt [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+    }
+
+    override fun layout(): Int = ar.com.wolox.android.R.layout.profile_fragment_page
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/ProfileFragment.kt]: Please remember to follow id's nomeclature: fragment_profile_page could be better

File: app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/news/INewsView.kt [hunk 11]
```diff
@@ -0,0 +1,6 @@
+package ar.com.wolox.android.example.ui.training.homePage.news
+
+interface INewsView {
+    fun addNews()
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/news/INewsView.kt]: Add an empty line above this

File: app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/news/INewsView.kt [hunk 12]
```diff
@@ -0,0 +1,6 @@
...
+    fun noNews()
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/news/INewsView.kt:12]: Add an empty line above here too

File: app/src/main/res/layout/fragment_viewpager.xml [hunk 13]
```diff
@@ -1,22 +1,61 @@
 <?xml version="1.0" encoding="utf-8"?>
-<LinearLayout
-    xmlns:android="http://schemas.android.com/apk/res/android"
+<android.support.constraint.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (lmiotti) [inline app/src/main/res/layout/fragment_viewpager.xml]: Send xmlns to a new line

File: app/src/main/res/layout/news_fragment_page.xml [hunk 14]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<android.support.constraint.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (lmiotti) [inline app/src/main/res/layout/news_fragment_page.xml]: Same as above

File: app/src/main/res/layout/news_fragment_page.xml [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="match_parent">
+
+        <android.support.v7.widget.RecyclerView
+            android:id="@+id/news_recycler"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_centerInParent="true"
+            tools:ignore="MissingConstraints" />
+    </android.support.v4.widget.SwipeRefreshLayout>
```
Comment (lmiotti) [inline app/src/main/res/layout/news_fragment_page.xml]: Add an empty line above this

File: app/src/main/res/layout/profile_fragment_page.xml [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+    <TextView
+        android:id="@+id/mTitle"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_centerInParent="true"
+        android:text="@string/profile_fragment"
+        tools:ignore="MissingConstraints" />
+</android.support.constraint.ConstraintLayout>
```
Comment (lmiotti) [inline app/src/main/res/layout/profile_fragment_page.xml:16]: Add an empty line above

File: app/src/main/res/layout/profile_fragment_page.xml [hunk 17]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<android.support.constraint.ConstraintLayout
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+    <TextView
```
Comment (lmiotti) [inline app/src/main/res/layout/profile_fragment_page.xml:8]: Add an empty line above

File: app/src/main/res/layout/recycler_news.xml [hunk 18]
```diff
@@ -0,0 +1,75 @@
+<?xml version="1.0" encoding="utf-8"?>
+<android.support.constraint.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (lmiotti) [inline app/src/main/res/layout/recycler_news.xml]: Send xmlns to a new line

File: app/src/main/res/layout/recycler_news.xml [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="88dp">
```
Comment (lmiotti) [inline app/src/main/res/layout/recycler_news.xml]: Solve this harcoded dimen

File: app/src/main/res/layout/recycler_news.xml [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:src="@drawable/ic_like_off"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintLeft_toRightOf="@id/vText_description"
+        app:layout_constraintRight_toRightOf="parent"
+        app:layout_constraintTop_toBottomOf="@id/vText_time"
+        tools:ignore="MissingConstraints" />
+
+
```
Comment (lmiotti) [inline app/src/main/res/layout/recycler_news.xml]: Remove this empty lines

File: app/src/main/res/values/dimens.xml [hunk 21]
```diff
@@ -44,6 +44,13 @@
     <dimen name="font_large_extra">20sp</dimen>
     <dimen name="font_big">22sp</dimen>
     <dimen name="font_huge">24sp</dimen>
+    <dimen name="image_view_height">180dp</dimen>
+    <dimen name="view_height">60dp</dimen>
+    <dimen name="margin">16dp</dimen>
+    <dimen name="button_height">40dp</dimen>
+    <dimen name="icon_size">40dp</dimen>
+    <dimen name="zero">0dp</dimen>
```
Comment (lmiotti) [inline app/src/main/res/values/dimens.xml]: Zero dimen is not really necesary, we made an exception for this case in layouts. Every custom dimen, please move it below "Add your custom dimens here"

File: app/src/main/res/values/strings.xml [hunk 22]
```diff
@@ -12,5 +12,30 @@
...
+    <string name="no_internet">No internet connection</string>
+    <string name="wolox">Wolox</string>
+    <string name="profile_fragment">Profile Fragment</string>
+    <string name="title">Title</string>
+    <string name="description">Description</string>
+    <string name="_15_min">15 min</string>
+    <string name="no_news">No more news to show</string>
+    <string name="profile">Profile</string>
+    <string name="news_fragment">News Fragment</string>
+    <string name="news">News</string>
```
Comment (lmiotti) [inline app/src/main/res/values/strings.xml:46]: I've made the same comment in other PRs, use more descriptive ids for strings, dimens, etc. For example, If the app is getting bigger I wouldn't know where no_news is called

**Review by lmiotti — CHANGES_REQUESTED** [review]
Please check the corrections we've made with Dylan. We need to beautify this code

File: app/build.gradle (hunk 5 above)
Comment (lmiotti) [reply app/build.gradle]: What happened with this?

File: app/src/main/java/ar/com/wolox/android/example/model/News.kt [hunk 23]
```diff
@@ -0,0 +1,29 @@
+package ar.com.wolox.android.example.model
+
+class News(cTitle: String, cText: String, cPicture: String) {
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/model/News.kt]: One important feature of Kotlin is removing getters and setters. Pleasee tell me if you need help with this

File: app/src/main/java/ar/com/wolox/android/example/model/User.kt [hunk 24]
```diff
@@ -0,0 +1,19 @@
+package ar.com.wolox.android.example.model
+
+import com.google.gson.annotations.SerializedName
+
+class User {
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/model/User.kt]: Same as News class

File: app/src/main/res/values/strings.xml (hunk 22 above)
Comment (lmiotti) [reply app/src/main/res/values/strings.xml:46]: What happened with this?

**Review by lmiotti — CHANGES_REQUESTED** [review]
There are some corrections that are missing, please check this again

File: app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/news/NewsFragment.kt (hunk 3 above)
Comment (lmiotti) [reply app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/news/NewsFragment.kt]: This is missing too!

File: app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/HomePageViewPagerFragment.kt [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+import ar.com.wolox.wolmo.core.adapter.viewpager.SimpleFragmentPagerAdapter
+import android.support.v4.util.Pair
+import ar.com.wolox.android.example.ui.training.homePage.news.NewsFragment
+
+/**
+ * LoginFragment is used for user to Login
+ */
+class HomePageViewPagerFragment @Inject constructor() : WolmoFragment<BasePresenter<Any>>() {
+
+    @Inject lateinit var mNewsFragment: NewsFragment
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/HomePageViewPagerFragment.kt]: Check this Ids, remember that in Kotlin we don't use m as a prefix

File: app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/news/INewsView.kt (hunk 12 above)
Comment (lmiotti) [reply app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/news/INewsView.kt:12]: What happened with this?

File: app/src/main/java/ar/com/wolox/android/example/ui/training/login/ILoginView.java [hunk 26]
```diff
@@ -0,0 +1,16 @@
+package ar.com.wolox.android.example.ui.training.login;
+
+/**
+ * IloginView is used for user to Login
+ */
+public interface ILoginView {
+
+    void loginSuccesful();
+    void incompletePassword();
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/ui/training/login/ILoginView.java:10]: Add empty lines between methods

**Review by lmiotti — CHANGES_REQUESTED** [review]
Some minor changes and we're almost done

**Review by lmiotti — APPROVED** [review]

**Comment by lmiotti** [comment]
Good job! We need @dylan-muszel revision and we can merge

File: app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/news/NewsFragment.kt [hunk 27]
```diff
@@ -0,0 +1,51 @@
...
+    private lateinit var viewAdapter: RecyclerView.Adapter<*>
+    private lateinit var viewManager: RecyclerView.LayoutManager
+    private lateinit var news: MutableList<News>
+
+    override fun layout(): Int = R.layout.fragment_news_page
+
+    override fun init() {
+        vSwipeRefreshLayout.setOnRefreshListener {
+            presenter.refreshNews()
+            vSwipeRefreshLayout.setRefreshing(false)
```
Comment (dylan-muszel) [inline app/src/main/java/ar/com/wolox/android/example/ui/training/homePage/news/NewsFragment.kt]: I only see this, but is important to know that the property access syntax in Kotlin is: `vSwipeRefreshLayout.isRefreshing = false`

**Review by dylan-muszel — CHANGES_REQUESTED** [review]
Resolve that little thing I commented and what we talked and I will be ok.

**Review by dylan-muszel — APPROVED** [review]

**Pull request merged**